src/main/java/by/custom/utilcalculator/domain/FileUserProgressStorage.java

```diff
@@ -0,0 +1,68 @@
+package by.custom.utilcalculator.domain;
+
+import by.custom.utilcalculator.exception.ReadingUserProgressFromFileException;
+import by.custom.utilcalculator.exception.UserFileNotFoundException;
+
+import java.io.*;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+public class FileUserProgressStorage implements IUserProgressStorage {
+    private final String databaseEnvVariable = System.getenv("CHATBOT_DATABASE_STORAGE");
+    private final String path = databaseEnvVariable + "%s.dat";
+
+    private static class FileUserProgressStorageHolder {
+        private static final FileUserProgressStorage FILE_USER_PROGRESS_STORAGE = new FileUserProgressStorage();
+    }
+
+    private FileUserProgressStorage() {
+        File databaseDirectory = new File(databaseEnvVariable);
+        if (!databaseDirectory.exists()) {
+            databaseDirectory.mkdirs();
+        }
+    }
+
+    public static FileUserProgressStorage getInstance() {
+        return FileUserProgressStorageHolder.FILE_USER_PROGRESS_STORAGE;
+    }
+
+    @Override
+    public void save(final UserProgress userProgress) {
+        try {
+            ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(buildFilePath(userProgress.getChatID())));
```

Comment:
Please, use [try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html)

Reply:
Done, thanks for reminding about this construction